### PR TITLE
Feature/juan/integration testing

### PR DESCRIPTION
**Pull request by juancoob**
**ISSUE**
#83 

**SUMMARY**
Add instrumentation tests to the app. Despite _FavoriteCharacterViewModelIntegrationTest_ and _CharacterDetailViewModelIntegrationTest_ are both not necessary, they were added just for practice purposes

### CODE REVIEW

File: app/src/androidTest/java/com/architectcoders/arquitectomarvel/fakeDataSources/FakeLocalDataSource.kt [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import kotlinx.coroutines.flow.flowOf
+import mockedCharacter
+import javax.inject.Inject
+
+class FakeLocalDataSource @Inject constructor() : LocalDataSource {
+
+    var fakeCharacterList: MutableList<Character> = mutableListOf()
+    var fakeFavoriteCharacterList: MutableList<FavoriteCharacterEntity> = mutableListOf()
+    var fakeComicList: MutableList<Comic> = mutableListOf()
+    var pwd: String = "pwd"
```
Comment (joseantoniocampillo) [inline app/src/androidTest/java/com/architectcoders/arquitectomarvel/fakeDataSources/FakeLocalDataSource.kt]: This value can be Null in production and it should be Nullable anyway for test

File: app/src/androidTest/java/com/architectcoders/arquitectomarvel/fakeDataSources/FakeLocalDataSource.kt [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    var hint: String = "hint"
```
Comment (joseantoniocampillo) [inline app/src/androidTest/java/com/architectcoders/arquitectomarvel/fakeDataSources/FakeLocalDataSource.kt]: (Idem)

File: app/src/androidTest/java/com/architectcoders/arquitectomarvel/fakeDataSources/FakeLocalDataSource.kt [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import com.architectcoders.data.source.LocalDataSource
+import com.architectcoders.domain.character.Character
+import com.architectcoders.domain.comic.Comic
+import kotlinx.coroutines.flow.flowOf
+import mockedCharacter
+import javax.inject.Inject
+
+class FakeLocalDataSource @Inject constructor() : LocalDataSource {
+
+    var fakeCharacterList: MutableList<Character> = mutableListOf()
```
Comment (joseantoniocampillo) [inline app/src/androidTest/java/com/architectcoders/arquitectomarvel/fakeDataSources/FakeLocalDataSource.kt]: I think you would have to fill a fakeList here and then interact with it through the methods of this implementation of the localdatasource ...

File: app/src/androidTest/java/com/architectcoders/arquitectomarvel/fakeDataSources/FakeLocalDataSource.kt [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    var fakeFavoriteCharacterList: MutableList<FavoriteCharacterEntity> = mutableListOf()
+    var fakeComicList: MutableList<Comic> = mutableListOf()
+    var pwd: String = "pwd"
+    var hint: String = "hint"
+
+    override suspend fun getLocalCharacterById(characterId: Int): Character = mockedCharacter
+
+    override suspend fun getLastTimeStampFromCharacterEntity(): Long = 1L
+
+    override suspend fun getLocalCharactersCount(): Int = 1
```
Comment (joseantoniocampillo) [inline app/src/androidTest/java/com/architectcoders/arquitectomarvel/fakeDataSources/FakeLocalDataSource.kt]: This must be:
 = fakeCharacterList.size

File: app/src/androidTest/java/com/architectcoders/arquitectomarvel/fakeDataSources/FakeLocalDataSource.kt [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+class FakeLocalDataSource @Inject constructor() : LocalDataSource {
+
+    var fakeCharacterList: MutableList<Character> = mutableListOf()
+    var fakeFavoriteCharacterList: MutableList<FavoriteCharacterEntity> = mutableListOf()
+    var fakeComicList: MutableList<Comic> = mutableListOf()
+    var pwd: String = "pwd"
+    var hint: String = "hint"
+
+    override suspend fun getLocalCharacterById(characterId: Int): Character = mockedCharacter
```
Comment (joseantoniocampillo) [inline app/src/androidTest/java/com/architectcoders/arquitectomarvel/fakeDataSources/FakeLocalDataSource.kt]: this must be:
```
override suspend fun getLocalCharacterById(characterId: Int): Character? =      
        fakeCharacterList.find { it.id == characterId }
```
So we need change in LocalDataSource interface for this in its function:
suspend fun getLocalCharacterById(characterId: Int): Character?
(Making it with a nullable return)

File: app/src/androidTest/java/com/architectcoders/arquitectomarvel/fakeDataSources/FakeLocalDataSource.kt [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+    override suspend fun isPasswordAlreadyStored(): Boolean = true
+
+    override suspend fun saveCredentials(password: String, recoveryHint: String) {
+        pwd = password
+        hint = recoveryHint
+    }
+
+    override suspend fun deleteCredentials() {
+        pwd = ""
```
Comment (joseantoniocampillo) [inline app/src/androidTest/java/com/architectcoders/arquitectomarvel/fakeDataSources/FakeLocalDataSource.kt]: I think the pwd and hint class variables should be set to null

File: app/src/androidTest/java/com/architectcoders/arquitectomarvel/fakeDataSources/FakeLocalDataSource.kt [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    var pwd: String = "pwd"
+    var hint: String = "hint"
+
+    override suspend fun getLocalCharacterById(characterId: Int): Character = mockedCharacter
+
+    override suspend fun getLastTimeStampFromCharacterEntity(): Long = 1L
+
+    override suspend fun getLocalCharactersCount(): Int = 1
+
+    override fun isLocalCharacterFavorite(characterId: Int): Any = flowOf(1)
```
Comment (joseantoniocampillo) [inline app/src/androidTest/java/com/architectcoders/arquitectomarvel/fakeDataSources/FakeLocalDataSource.kt]: maybe better:
```
flowOf(
    fakeFavoriteCharacterList.find { it.id == characterId }?:0
)
```

File: app/src/androidTest/java/com/architectcoders/arquitectomarvel/integration/MainViewModelIntegrationTest.kt [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+        mainViewModel.saveCredentials("myPassword", "myHint")
+        mainViewModel.checkIfHintIsCorrect("myHint") {
+            assertTrue(it)
+        }
+    }
+
+    @Test
+    fun ifDeleteCredentials_checkIfPasswordIsEmpty() {
+        mainViewModel.deleteCredentials()
+        mainViewModel.checkIfPasswordIsCorrect("") {
```
Comment (joseantoniocampillo) [inline app/src/androidTest/java/com/architectcoders/arquitectomarvel/integration/MainViewModelIntegrationTest.kt]: It does not correspond to production.
deleteCredentials, implies that checkIfPasswordIsCorrect ("") is currently crashing the app. Review it.

**Review by joseantoniocampillo — CHANGES_REQUESTED** [review]
For integration tests I would consider providing an in-memory roomdatabase with:
Room.inMemoryDatabaseBuilder.

**Comment by juancoob** [comment]
> For integration tests I would consider providing an in-memory roomdatabase with: Room.inMemoryDatabaseBuilder.

Hello @joseantoniocampillo I'll use your approach by removing fakes to use real data. Thanks! :smile:

File: app/src/androidTest/java/com/architectcoders/arquitectomarvel/fakeDataSources/FakeLocalDataSource.kt (hunk 1 above)
Comment (juancoob) [reply app/src/androidTest/java/com/architectcoders/arquitectomarvel/fakeDataSources/FakeLocalDataSource.kt:21]: You are right @joseantoniocampillo ! I'll change it :smile:

**Review by jujoru — APPROVED** [review]
Ok

**Review by JorgeGreek — APPROVED** [review]
ok

**Review by joseantoniocampillo — APPROVED** [review]
That's good

**Pull request merged**